pkg/diagnose/runner.go

```diff
@@ -407,45 +435,88 @@ func runStdOut(w io.Writer, diagCfg diagnosis.Config, run func(diagnosis.Config)
 		color.NoColor = true
 	}
 
-	fmt.Fprintf(w, "=== Starting diagnose ===\n")
+	if !diagCfg.JSONOutput {
+		fmt.Fprintf(w, "=== Starting diagnose ===\n")
+	}
 
 	diagnoses, err := run(diagCfg)
-	if err != nil && !diagCfg.RunLocal {
-		fmt.Fprintln(w, color.YellowString(fmt.Sprintf("Error running diagnose in Agent process: %s", err)))
-		fmt.Fprintln(w, "Running diagnose command locally (may take extra time to run checks locally) ...")
-
-		// attempt to do so locally
+	if err != nil {
+		errMsg := fmt.Sprintf("Error running diagnose in Agent process: %s", err)
+		action := "Running diagnose command locally (may take extra time to run checks locally) ..."
+		if diagCfg.JSONOutput {
+			errJSON, _ := json.MarshalIndent(map[string]string{
+				"error":  errMsg,
+				"action": action,
+			}, "", "  ")
+			fmt.Fprintln(w, string(errJSON))
+		} else if !diagCfg.RunLocal {
+			fmt.Fprintln(w, color.YellowString(errMsg))
+			fmt.Fprintln(w, action)
+		}
+		// Attempt to run diagnose locally
 		diagCfg.RunLocal = true
 		diagnoses, err = run(diagCfg)
 	}
 
 	if err != nil {
-		fmt.Fprintln(w, color.RedString(fmt.Sprintf("Error running diagnose: %s", err)))
+		if diagCfg.JSONOutput {
+			errJSON, _ := json.MarshalIndent(map[string]string{
+				"error": "Error running diagnose locally: " + err.Error(),
+			}, "", "  ")
+			fmt.Fprintln(w, string(errJSON))
+		} else {
+			fmt.Fprintln(w, color.RedString(fmt.Sprintf("Error running diagnose: %s", err)))
+		}
 		return err
 	}
 
-	var c counters
+	var diagnosesJSON []diagnosis.DiagnosesJSON
+	if diagCfg.JSONOutput {
+		diagnosesJSON = []diagnosis.DiagnosesJSON{}
+	}
 
+	var c counters
 	lastDot := false
 	for _, ds := range diagnoses {
 		suiteAlreadyReported := false
+		var diagnosesSuiteJSON diagnosis.DiagnosesJSON
+		if diagCfg.JSONOutput {
```

Comment:
This function contains a lot of `if diagCfg.JSONOutput` which makes the code a little harder to read.
As there are two possible outputs, I would suggest defining a `DiagnoseWriter` interface
```go
type DiagnoseWriter interface {
   WriteError(string)
   WriteText(string) // Write text such as `"Running diagnose command locally (may take extra time to run checks locally) ..."`
   WriteDiagnosisResult()
   ...
}
```
Then you can easily have 2 implementations for this interface, one for stdout and one for JSON.
The benefit of this change is:
- You remove some complexity from this function
- You can easily add a new output if needed.
What do you think?